README.md
# TechPack_0siribix V2.06

TechPack, a Mining, Crafting, &amp; Farming Modpack for Minetest.</br>
Created by joe7575 with extensions by 0siribix to be compatible with Drawers_0siribix,</br>
  add refill ability, and pulling stacks from anything that doesn't have a HighPerf variant.

![TechPack](https://github.com/0siribix/techpack_0siribix/blob/master/screenshot.png)

**After update to v2, don't forget to activate the new mods "techpack_stairway" and "techpack_warehouse" as part of the mod pack.**


TechPack is a collection of following Mods:

* tubelib, a Mod for item exchange via lumber tubes and wireless message communication between nodes.
* tubelib_addons1, a Tubelib extension with mining, farming, and crafting nodes
* tubelib_addons2, a Tubelib extension with control nodes
* tubelib_addons3, a Tubelib extension with high performance nodes
* techpack_stairway, Ladders, stairways, and bridges for your machines
* techpack_warehouse, Warehouse Boxes in steel, copper, and gold
* gravelsieve, a Mod to sieve ores from gravel.
* smartline, a Mod with small and smart sensors, actors and controllers.
* safer_lua, a subset of the language Lua for safe and secure Lua sandboxes
* SaferLua Controller - a controller to be programmed in LUA
* lcdlib - a display lib used by smartline

**A TechPack Tutorial is available as ![Wiki](https://github.com/0siribix/techpack_0siribix/wiki)**

**Machine data is stored in memory and not in the nodes itself.**</br>
**Therefore, never move or copy machines or tubes by means of Worldedit.**</br>
**The moved/copied nodes will not have valid node numbers, which could lead to a server crash.**

TechPack provides:

- lumber tubes to connect 2 nodes
- a Pusher node to pull/push items through tubes
- a Distributor node with 4 output channels to sort incoming items
- a Blackhole node which lets all items disappear
- Button/switches to send "switch on/off" messages
- a Forceload block to keep your machines operational
- Several lamp nodes in different colors (can be switched on/off)
- a Quarry node to dig for stones and other ground nodes
- a Harvester node to chop wood, leaves, flowers and crops
- a Grinder node to grind all kinds of cobblestone to gravel, gravel to sand, and sand to clay
- a Gravelsieve node to sieve ores from gravel
- an Autocrafter node for automated crafting of tools and items
- a Fermenter node to produce Bio Gas from leaves
- a Reformer node to produce Bio Fuel from Bio Gas (the Bio Fuel is needed by Harvester and Quarry nodes)
- a Funnel node to collect dropped items
- two Timer nodes for a daytime controlled sending of commands (on/off)
- two Sequencer nodes for a waiting time controlled sending of commands (on/off)
- an item Detector node sending commands (on/off)
- a Repeater node to distribute received commands to connected nodes
- a Logic Not node to invert on/off commands
- Gate/Door nodes in different textures to be controlled via on/off commands
- an Access Lock node with number key field
- a Mesecon Converter node to translate tubelib commands in mesecon commands and vice versa
- a Programmer tool to simply collect node numbers
- a Player Detector node
- a Controller node with "IF this then that" rules, which allows:
  - reading node states
  - receiving commands from other nodes
  - sending commands and alarms
  - sending mails or chat messages
  - output on a display
- a Controller to be programmed in Lua
- a Terminal to be connected to the Lua Controller
- a Signal Tower node showing machine states
- a Display node for text outputs of the Controller
- Metal ladders, stairways, and bridges
- Warehouse Boxes in steel, copper, and gold
- A chest cart for the mod minecart


TechPack supports the following mods:

- Farming Redo (Harvester, Fermenter)</br>
- Ethereal (Harvester, Quarry, Fermenter)</br>
- Pipeworks (Gravel Sieve)</br>
- Hopper (Gravel Sieve)</br>
- Mesecon (Mesecon Converter)</br>
- Drawers_0siribix


### Configuration

The following can be changed in the minetest menu (Settings -> Advanced Settings -> Mods -> tubelib) or directly in 'minetest.conf'
- Maximum number of Forceload Blocks per player
- Enable Basalt Stone (and disable ore generation via Cobblestone generator)
- Machine aging value to calculate the lifetime of machines

Example for 'minetest.conf':

```LUA
tubelib_basalt_stone_enabled = false
tubelib_max_num_forceload_blocks = 12
tubelib_machine_aging_value = 200
```

Example for a v1 compatible 'minetest.conf':

```LUA
tubelib_basalt_stone_enabled = false
tubelib_max_num_forceload_blocks = 0
tubelib_machine_aging_value = 999999
```


#### Maximum number of Forceload Blocks per player

Default value is 12.</br>
A higher number allows to build larger farms and machines which keep loaded, but increases the server load, too.</br>
But the areas are only loaded when the player is online.</br>
To be able to use e.g. 12 forceloaded blocks per player, the pararamter 'max_forceloaded_blocks' in 'minetest.conf' has to be ajusted.</br>


#### Enable Basalt Stone (and disable ore generation via Cobblestone generator)

The lava/water Cobblestone generator allows to produce infinite Cobblestone. By means of Quarry,</br>
Grinder, and Gravel Sieve it allows to infinite generate ores.</br>
This can be disabled by means of the setting parameter. If enabled, the Cobblestone</br>
generator generates Basalt instead, which only can be used for building purposes.


#### Disable machines of absent players

Machines will defect when the owner has not been online for 7 days (configurable)

### License

Copyright (C) 2017-2021 Joachim Stolberg</br>
Code: Licensed under the GNU AGPL version 3 or later. See LICENSE.txt</br>
Textures: CC BY-SA 3.0


## Credits


### Contributors

- oversword (PR #43, #57, #58, #59, #60, #62, #68, #74, #76, and many more)
- afkplayer5000 (PR #70, #71)
- andrenete (PR #37, #66)
- fluxionary (PR #27, #28, #30, #31, #34, #54)
- Arigatas (PR #51, #53)
- realmicu (PR #6, #8, #12)
- theFox6 (PR #3, #4)
- superfloh247 (PR #89, #88, #87)
- SciFurz (via forum)


### Dependencies

default, doors, intllib, basic_materials</br>
tubelib2 (![GitHub](https://github.com/joe7575/tubelib2))</br>
Tubelib Color Lamps optional: unifieddyes</br>
SmartLine Controller optional: mail</br>
Gravelsieve optional: moreores, hopper, pipeworks</br>
tubelib_addons1 optional: unified_inventory</br>
tubelib_addons3 optional: minecart


### History

- 2018-03-18  V1.00  * Tubelib, tubelib_addons1, tubelib_addons2, smartline, and gravelsieve combined to one modpack.
- 2018-03-24  V1.01  * Support for Ethereal added
- 2018-03-27  V1.02  * Timer improvements for unloaded areas
- 2018-03-29  V1.03  * Area protected chest added to tubelib_addons1
- 2018-03-31  V1.04  * Maintenance, minor issues, Unifieddyes support for Color Lamp, Street Lamp added
- 2018-04-27  V1.05  * Ceiling lamp added, further improvements
- 2018-06-09  V1.06  * Recipes with API to grinder added
- 2018-06-17  V1.07  * Tube placement completely reworked
- 2018-06-22  V1.08  * Lua Controller and SaferLua added
- 2018-07-22  V1.09  * Item counters for Pusher/Distributor and tubelib commands for Autocrafter added
- 2018-08-08  V1.10  * tubelib_addon3 with high performance nodes added
- 2018-08-13  V1.11  * Detector node added
- 2018-08-14  V1.12  * Teleporter node added
- 2018-08-28  V1.13  * Smartline Controller completely revised. Liquid Sampler added
- 2018-09-10  V1.14  * Distributor performance improved, chest commands added
- 2018-09-15  V1.15  * Smartline Controller command added, chest commands improved
- 2018-09-20  V1.16  * Switched from "Display mod pack" to lcdlib
- 2018-12-23  V2.xx  * on the way to v2
- 2018-12-29  V2.00  * beta
- 2019-01-12  V2.00  * release
- 2019-01-27  V2.01  * SaferLua Controller Terminal added
- 2019-01-28  V2.02  * Logic Not added, output reduction on Harvester, Fermenter, and Gravel Sieve
- 2019-04-23  V2.03  * Piston/WorldEdit/replacer detection added, farming and grinder recipes added
- 2020-11-20  V2.04  * Switch to AGPL v3, adapt to minetest 5.3, add translation support, fix minor bugs
- 2021-01-24  V2.05  * PR #74, #76: Implement checks for valid connection sides for many nodes
- 2021-06-06  V2.06  * PR #78 - #89, chest cart added
- 2021-09-03  V2.07  * FR #103, Add Altitude to harvester menu


## New in v2 (from players point of view)

- Almost all machines break after a certain amount of time (switch into the state 'defect') and have to be repaired.
- A Repair Kit is available to repair defect machines.
- A Forceload block (16x16x16) is added which keeps the corresponding area loaded and the machines operational as far as the player is logged in.
- The Quarry now uses LVM techniques to go down up to 100 meter
- Ladders, stairways, and bridges added for the machines (techpack_stairway)
- Industrial lamps
- Warehouse Boxes in steel, copper, and gold for your warehouse/stock (techpack_warehouse)


## New in v2 (from admins point of view)

- settingtypes introduced with the following settings: tubelib_max_num_forceload_blocks, tubelib_basalt_stone_enabled, tubelib_machine_aging_value
- the new mods 'techpack_stairway' and 'techpack_warehouse' have to be enabled
- TechPack depends now on the mod 'basic_materials' and 'tubelib2' (![GitHub](https://github.com/joe7575/tubelib2))

See ![releasenotes.txt](https://github.com/0siribix/techpack_0siribix/blob/master/releasenotes.md) for further information


releasenotes.md
# Release Notes for ModPack TechPack [techpack]

## Extensions by 0siribix (2022-01-25)
- Add vacuum and HP vacuum
- Add support for drawers_0siribix
- Add refill code (pushing nodes will now push a partial stack to refill if there is room)
- Added ability to pull stacks from all nodes which don't have a HighPerf variant

## V2.06.00 (2021-06-06)

### Additions
- Add chest cart for the mod minecart (optional)
- Add support for the mod "underch" (#88)

### Removals

### Changes
- Account for new red & yellow pepper varieties (#86)
- Pusher improvements (#84, #83)

### Fixes
- Fix crash with "ethereal:strawberry" (#89)


## V2.05.00 (2021-01-24)

### Additions

### Removals

### Changes
- Pushers only connect from the left & right
- Detectors only connect from the left & right, still need condition to stop pushing in from the right
- Funnels refuse to connect from the top
- Liquid samplers & quarries refuse to connect from the front (left)
- Black hole push filter replaced with valid_sides logic

### Fixes
- Pushers can be used to send items into the invalid faces of a node (#72)


## V2.04.01 (2020-12-18)

### Additions

### Removals

### Changes
- Suggestion to check replanting against commonly plantable nodes
  instead of quarry-able GroundNodes (pull request #60 from oversword)
- Add a on_blast callback to tubes so they update after being destroyed
  (pull request #58 from oversword)
- Disallow non-fuel items to be accepted as fuel quarries and harvesters
  (pull request #57 from oversword)
- Check harvester protection for each node it attempts to harvest
  (pull request #59 from oversword)

### Fixes
- Warehouse voiding items when both pushing in and pulling out
  (issue #61, fixed by oversword)



## V2.04.00 (2020-11-20)

### Additions
- Add minetest.translator support (issue #56)
- Add German translation
- Add new minetest v5 nodes

### Removals

### Changes
- Switch from LGPL to AGPL v3

### Fixes
- tubelib_addons2:doorblock works incorrectly (issue #55)



## V2.03.12 (2020-07-23)

### Additions

### Removals

### Changes

### Fixes
- string.split bugfix (issue #49)


## V2.03.11 (2020-07-21)

### Additions

### Removals

### Changes

### Fixes
- SaferLua Controller comments bugfix (issue #48)
- SmartLine Controller bugfix


## V2.03.10 (2020-06-06)

### Additions

### Removals

### Changes
Merged fork from BlockySurvival back to base (pull request #46)

### Fixes



## V2.03.09 (2020-03-23)

### Additions

### Removals

### Changes
Tubelib database optimized to reduce the size of the mod storage file (based on issue #41)

### Fixes



## V2.03.08 (2020-03-03)

### Additions

### Removals

### Changes

### Fixes
Set Idle state instead of fault when there is no liquid (issue #44)



## V2.03.07 (2019-09-14)

### Additions

### Removals

### Changes

### Fixes
- Stack overflow bugfix for the detector from fluxionary (issue #39)



## V2.03.06 (2019-08-08)

### Additions

### Removals

### Changes

### Fixes
- Fixes to distributor metadata_inventory functions



## V2.03.05 (2019-08-06)

### Additions
- Tubelib_addons3 "HighPerf Funnel" added (see isssue #36)

### Removals

### Changes

### Fixes



## V2.03.04 (2019-07-20)

### Additions

### Removals

### Changes

### Fixes
- Use Unified Dyes on_dig where needed (issue #35)


## V2.03.03 (2019-05-22)

### Additions

### Removals

### Changes

### Fixes
- Bug fixes from obl3pplifp (Blocky Survival) (issues #27, #28)
- Unstackable items vanish when using HighPerf pushers (issue #29)


## V2.03.02 (2019-05-09)

### Additions

### Removals

### Changes
- HighPerf Pusher support for autocrafter, grinder, and harvester
  added (issue #22, #23)
- Both distributor behaviours changed (issue #26)

### Fixes
- Black Hole "items disappeared" counter bugfix (issue #24)
- HighPerf distributor behaviour without an active filtered channel
  changed (issue #25)



## V2.03.01 (2019-05-03)

### Additions

### Removals

### Changes

### Fixes
- Warehouse nodes inventory handling bugfix



## V2.03 (2019-04-23)

### Additions
- Farming and Grinder recipes added (thanks to obl3pplifp)
- Support for Signs Bot chest and Box added

### Removals

### Changes

### Fixes
- Piston/WorldEdit/Replacer tool detection added. If node is replaced/removed/copied,
  it will switch to a "defect" dummy node.



## V2.02.06 (2019-03-23)

### Additions

### Removals

### Changes

### Fixes
- Warehouse boxes bugfix for the Blackhole bugfix



## V2.02.05 (2019-03-14)

### Additions

### Removals

### Changes

### Fixes
- Blackhole tube side bugfix



## V2.02.04 (2019-03-02)

### Additions

### Removals

### Changes

### Fixes
- Fermenter leaves bugfix (moretrees)



## V2.02.03 (2019-02-27)

### Additions

### Removals

### Changes

### Fixes
- industriallamp bugfix



## V2.02.02 (2019-02-16)

### Additions
- sl_controller: add battery status command

### Removals

### Changes

### Fixes
- forceload block bugfix



## V2.02.01 (2019-01-29)

### Additions

### Removals

### Changes
- Terminal related Controller commands changed

### Fixes
- Terminal output bugfix
- Controller event handling bugfix



## V2.02 (2019-01-28)

### Additions
- Tubelib_addons2 "Logic Not" added
- second (classical) SaferLua Terminal added

### Removals

### Changes
- Output reduction on Harvester (cycle time from 4 to 6 s),
  Fermenter (from 2 to 3 input items needed per bio gas),
  and Gravel Sieve (rarity from 1 to 1.5)

### Fixes
- removing the Gate block returns the original block


## V2.01 (2019-01-27)

### Additions
- SaferLua Terminal to be connected to the SaferLua Controller

### Removals

### Changes
- Server: formspec added to enter valid usernames for server access

### Fixes



## V2.00.07 (2019-01-27)

### Additions

### Removals

### Changes
- SmartLine Controller: Own node number to on/off command added to be able
  to cascade SmartLine Controllers

### Fixes



## V2.00.06 (2019-01-22)

### Additions

### Removals

### Changes

### Fixes

tubelib/init.lua
--[[

	Tube Library
	============

	Copyright (C) 2017-2020 Joachim Stolberg

	AGPL v3
	See LICENSE.txt for more information

	History:
	2017-09-08  v0.01  first version
	2017-09-12  v0.02  bugfix in tubelib.get_pos() and others
	2017-09-21  v0.03  function get_num_items added
	2017-09-26  v0.04  param side added, node blackhole added
	2017-10-06  v0.05  Parameter 'player_name' added, furnace fuel detection changed
	2017-10-08  v0.06  tubelib.get_node_number() added, tubelib.version added
	2017-10-29  v0.07  Pusher bugfix, commands start/stop replaced by on/off
	2017-11-02  v0.08  Data base changed, aging of node positions added
	2017-11-04  v0.09  functions set_data/get_data added
	2018-01-27  v0.10  WLAN Chip added, recipes reviced, Pusher state 'blocked' added,
                       function send_request changed
	2018-08-09  v1.00  Extracted from TechPack to be used as standalone mod
	                   - new tubing algorithm
	                   - tubelib.pull_stack()/tubelib.get_stack() added
	                   - item counter for pusher/distributor added
	2018-12-21  v2.00  Switch to tubelib2, "defect" nodes and "Repair Kit" added,
                       Forceload block added, Basalt as Cobble Stone alternative added

]]--


tubelib = {
	NodeDef = {},		-- node registration info
}

-- Load support for I18n
tubelib.S = minetest.get_translator("tubelib")
local S = tubelib.S

tubelib.version = 2.00

tubelib.max_num_forceload_blocks = tonumber(minetest.setting_get("tubelib_max_num_forceload_blocks")) or 10
tubelib.basalt_stone_enabled = minetest.setting_get("tubelib_basalt_stone_enabled") == "true"
tubelib.machine_aging_value = tonumber(minetest.setting_get("tubelib_machine_aging_value")) or 100
tubelib.refill = true


--------------------------- conversion to v0.04
minetest.register_lbm({
	label = "[Tubelib] Distributor update",
	name = "tubelib:update",
	nodenames = {"tubelib:distributor", "tubelib:distributor_active"},
	run_at_every_load = false,
	action = function(pos, node)
		local meta = minetest.get_meta(pos)
		if minetest.deserialize(meta:get_string("filter")) == nil then
			local filter = {false,false,false,false}
			meta:set_string("filter", minetest.serialize(filter))
		end
		local inv = meta:get_inventory()
		inv:set_size('yellow', 6)
		inv:set_size('green', 6)
		inv:set_size('red', 6)
		inv:set_size('blue', 6)
	end
})

--------------------------- conversion to v0.10
minetest.register_lbm({
	label = "[Tubelib] Button update",
	name = "tubelib:update2",
	nodenames = {"tubelib:button", "tubelib:button_active"},
	run_at_every_load = false,
	action = function(pos, node)
		local meta = minetest.get_meta(pos)
		local number = meta:get_string("number")
		if number ~= "" then
			meta:set_string("numbers", number)
			meta:set_string("number", nil)
		end
	end
})

minetest.register_craftitem("tubelib:wlanchip", {
	description = S("Tubelib WLAN Chip"),
	inventory_image = "tubelib_wlanchip.png",
})


minetest.register_craft({
	output = "tubelib:wlanchip 8",
	recipe = {
		{"default:mese_crystal", "default:copper_ingot", ""},
		{"default:gold_ingot",   "default:glass",	""},
		{"", "", ""},
	},
})


if tubelib2.version < 2.0 then
	error("TechPack/Tubelib requires tubelib2 version 2.0 or newer!!!")
else
	dofile(minetest.get_modpath("tubelib") .. "/tubes.lua")
	dofile(minetest.get_modpath("tubelib") .. "/command.lua")
	-- conversion from v1.16 to v2.00
	dofile(minetest.get_modpath("tubelib") .. "/migrate.lua")
	dofile(minetest.get_modpath("tubelib") .. "/states.lua")
	dofile(minetest.get_modpath("tubelib") .. "/defect.lua")
	dofile(minetest.get_modpath("tubelib") .. "/node_states.lua")
	dofile(minetest.get_modpath("tubelib") .. "/pusher.lua")
	dofile(minetest.get_modpath("tubelib") .. "/blackhole.lua")
	dofile(minetest.get_modpath("tubelib") .. "/button.lua")
	dofile(minetest.get_modpath("tubelib") .. "/lamp.lua")
	dofile(minetest.get_modpath("tubelib") .. "/distributor.lua")
	dofile(minetest.get_modpath("tubelib") .. "/legacy_nodes.lua")
	dofile(minetest.get_modpath("tubelib") .. "/repairkit.lua")
	dofile(minetest.get_modpath("tubelib") .. "/mark.lua")
	dofile(minetest.get_modpath("tubelib") .. "/forceload.lua")
	dofile(minetest.get_modpath("tubelib") .. "/basalt.lua")
end
